### PR TITLE
AIP-51 - Executor Coupling in Logging

### LINKED ISSUE #27931: AIP-51 - Executor Coupling in Logging
### Overview
The FileTaskHandler logging class has a special branch for Kubernetes, to fetch logs from the pod (Note: this is currently hardcoded to KubernetesExecutor so this would fail to run for the CeleryKubernetesExecutor or LocalKubernetesExecutor).
### Examples
- **4a**) See code: https://github.com/apache/airflow/blob/27e2101f6ee5567b2843cbccf1dca0b0e7c96186/airflow/utils/log/file_task_handler.py#L157-L192
### Proposal
Adding a `get_task_logs(ti: TaskInstance | None)` method on the `BaseExecutor` class, which Executor subclasses could implement is one viable approach, since other Executors could certainly make use of such a paradigm, particularly containerized or cloud based executors.

### PR DESCRIPTION
**Pull request by snjypl**
Fixes: https://github.com/apache/airflow/issues/27931

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/utils/log/file_task_handler.py
Comment (o-nikolas) [inline airflow/utils/log/file_task_handler.py:251]: Previously in the exception/end of log cases the `log` as well as `{"end_of_log": True}` was returned. So shouldn't you return the whole tuple here?

File: airflow/executors/base_executor.py
Comment (o-nikolas) [inline airflow/executors/base_executor.py]: This code to fetch logs from workers over the network is not related to executors, so I don't think this implementation belongs in the base executor. I'd actually propose to put this back into `file_task_handler.py` and for this method on the base executor to just be a stub method with a `pass`.

**Review by o-nikolas — CHANGES_REQUESTED** [review]

File: airflow/utils/log/file_task_handler.py
Comment (snjypl) [reply airflow/utils/log/file_task_handler.py:251]: task_log  is the returned tuple | str.  incase of exception, we are returning the tuple

File: airflow/executors/base_executor.py
Comment (snjypl) [reply airflow/executors/base_executor.py]: i was thinking, getting the log from network  would be the default base executor behavior and getting from the pod a special case for kube executor.   will put this code back in file_task_handler.

Comment (o-nikolas) [reply airflow/executors/base_executor.py]: Perfect, thanks!

File: airflow/utils/log/file_task_handler.py
Comment (o-nikolas) [inline airflow/utils/log/file_task_handler.py]:
```suggestion
            log += f"*** Local log file does not exist, trying to fetch logs from executor environment ***\n\n"
```

This more closely matches what was there previously as well as the new context you added.

File: airflow/utils/log/file_task_handler.py
Comment (o-nikolas) [inline airflow/utils/log/file_task_handler.py:248]: Should add a log here, saying that we're now falling back to fetching from worker.

File: airflow/utils/log/file_task_handler.py
Comment (o-nikolas) [inline airflow/utils/log/file_task_handler.py:239]: You may need to first check if that method exists on the executor class before calling it or update some of the other executors, pending the result of the discussion here: https://github.com/apache/airflow/issues/28276#issuecomment-1344899475

File: tests/utils/test_log_handlers.py
Comment (o-nikolas) [inline tests/utils/test_log_handlers.py:377]: You've nicely refactored `FileTaskHandler._read` to be unittestable. You can mock `os.path.exists(location)` to return false and also mock the kubernetes executor, then ensure `get_task_log` was called once with the expected ti input. You should then swap the executor to one that doesn't have an implementation and ensure you get None back (you shouldn't need to mock in that case since it has no implementation) and that the `_get_task_log_from_worker` method is called once (will need to mock that one).

**Review by o-nikolas — CHANGES_REQUESTED** [review]

**Review by o-nikolas — CHANGES_REQUESTED** [review]
Marking my review as request for changes regarding unit testing (see [here](https://github.com/apache/airflow/pull/28161/files#r1044921868))

**Comment by snjypl** [comment]
> Marking my review as request for changes regarding unit testing (see [here](https://github.com/apache/airflow/pull/28161/files#r1044921868))
@o-nikolas  i have added some unittests,  please take a look when you get a chance.  Thanks.

File: airflow/executors/base_executor.py
Comment (o-nikolas) [inline airflow/executors/base_executor.py:357]: Just noticed that this is missing a doc string. Do you mind adding one that explains what this method does and that expectation is child classes implement it (if desired) and also documenting the params/return values.

**Review by o-nikolas — CHANGES_REQUESTED** [review]
Fantastic unit tests! Thanks for adding those.

Just left one other small request for a doc string.


Also, are you able to do any UAT/manual testing to ensure things are working as expected as compared to before this change?

File: airflow/executors/base_executor.py
Comment (snjypl) [reply airflow/executors/base_executor.py:357]: Done

**Comment by snjypl** [comment]
> Fantastic unit tests! Thanks for adding those.
> 
> Just left one other small request for a doc string.
> 
> Also, are you able to do any UAT/manual testing to ensure things are working as expected as compared to before this change?

@o-nikolas  i did some manual testing with  different executors. it is working as expected. 

in main there is an issue with  `LocalKubernetesExecutor`. scheduler is not serving the logs for local task. https://github.com/apache/airflow/pull/28638 . i tested with a patched version, and it is working fine. 



`celery executor` 
```
airflow-worker-0.airflow-worker.airflow.svc.cluster.local
*** Local log file does not exist: /opt/airflow/logs/dag_id=example_bash_operator/run_id=manual__2022-12-29T19:06:34.783246+00:00/task_id=runme_0/attempt=1.log
*** Failed to fetch log from executor. Falling back to fetching log from worker.
*** Fetching from: http://airflow-worker-0.airflow-worker.airflow.svc.cluster.local:8793/log/dag_id=example_bash_operator/run_id=manual__2022-12-29T19:06:34.783246+00:00/task_id=runme_0/attempt=1.log

[2022-12-29, 19:06:37 UTC] {taskinstance.py:1084} INFO - Dependencies all met for <TaskInstance: example_bash_operator.runme_0 manual__2022-12-29T19:06:34.783246+00:00 [queued]>
[2022-12-29, 19:06:37 UTC] {taskinstance.py:1084} INFO - Dependencies all met for <TaskInstance: example_bash_operator.runme_0 manual__2022-12-29T19:06:34.783246+00:00 [queued]>
[2022-12-29, 19:06:37 UTC] {taskinstance.py:1282} INFO - 
```

`LocalKubernetesExecutor ` local task
```
airflow-scheduler-0.airflow-scheduler.airflow.svc.cluster.local
*** Local log file does not exist: /opt/airflow/logs/dag_id=example_local_kubernetes_executor/run_id=manual__2022-12-29T19:11:14.004608+00:00/task_id=task_with_local_executor/attempt=1.log
*** Failed to fetch log from executor. Falling back to fetching log from worker.
*** Fetching from: http://airflow-scheduler-0.airflow-scheduler.airflow.svc.cluster.local:8793/log/dag_id=example_local_kubernetes_executor/run_id=manual__2022-12-29T19:11:14.004608+00:00/task_id=task_with_local_executor/attempt=1.log

[2022-12-29, 19:11:15 UTC] {taskinstance.py:1084} INFO - Dependencies all met for <TaskInstance: example_local_kubernetes_executor.task_with_local_executor manual__2022-12-29T19:11:14.004608+00:00 [queued]>
```
`LocalKubernetesExecutor ` k8s task

```
example-local-kubernetes-executor-task-with-kubernetes-q2qf6dpa
*** Local log file does not exist: /opt/airflow/logs/dag_id=example_local_kubernetes_executor/run_id=manual__2022-12-29T19:11:14.004608+00:00/task_id=task_with_kubernetes_executor/attempt=1.log
*** Trying to get logs (last 100 lines) from worker pod example-local-kubernetes-executor-task-with-kubernetes-q2qf6dpa ***


/home/airflow/.local/lib/python3.7/site-packages/airflow/cli/cli_parser.py:925 DeprecationWarning: The namespace option in [kubernetes] has been moved to the namespace option in [kubernetes_executor] - the old setting has been used, but please update your config.
[2022-12-29, 19:11:30 UTC] {dagbag.py:538} INFO - Filling up the DagBag from /opt/airflow/dags/example_local_kubernetes_executor.py
```

**Comment by o-nikolas** [comment]
> @o-nikolas i did some manual testing with different executors. it is working as expected.

Fantastic! Thanks @snjypl!

Sorry for the delayed response but I was out on holiday for the past couple weeks, just catching back up now.

Looks like there are some conflicts in main that you need to rebase and resolve, otherwise it LGTM!

**Comment by snjypl** [comment]
@o-nikolas   i have resolved the merge conflicts.

File: airflow/executors/kubernetes_executor.py
Comment (o-nikolas) [inline airflow/executors/kubernetes_executor.py:782]: Do we really want to suppress any and all exceptions here? What is the specific error case this is covering?

Comment (snjypl) [reply airflow/executors/kubernetes_executor.py:782]: @o-nikolas  those changes were part of https://github.com/apache/airflow/pull/28546 . i simply moved it from `file_task_handler` to `kubernetes_executor` while resolving merge conflicts. 

@dstandish  might be able to help with it.

Comment (o-nikolas) [reply airflow/executors/kubernetes_executor.py:782]: I see, fair enough

**Review by o-nikolas — APPROVED** [review]
Looks like the branch needs rebasing, but otherwise looks good!
Update: Actually you've merged above, looks good, just need a second approval for a core change like this :) (CC: @pierrejeambrun)

File: airflow/executors/kubernetes_executor.py
Comment (dstandish) [reply airflow/executors/kubernetes_executor.py:782]: It's just a simpler way to do this https://github.com/apache/airflow/pull/28546/files#diff-e7f34f73940eb52d92bb991abedc1c963431c5373c12dff739c8fb7d03e93d3aL195

**Comment by o-nikolas** [comment]
Anyone have some time to give this a second review/approval? Would be nice to get this merged for @snjypl
Maybe @potiuk, @eladkal or @pierrejeambrun?

**Comment by snjypl** [comment]
@potiuk @eladkal @pierrejeambrun  will be great if you could review this PR whenever you get a chance !

File: airflow/executors/kubernetes_executor.py
Comment (eladkal) [inline airflow/executors/kubernetes_executor.py:810]: Not really part of this PR but feels like the right place to ask.

Why do we raise these exceptions and not write the issue to the log and return it? (Like lines 285-287)

```
        except Exception as f:
            log += f"*** Unable to fetch logs from worker pod {ti.hostname} ***\n{str(f)}\n\n"
            return log, {"end_of_log": True}
```

I wonder if this is the reason users sometimes don't see the task log and it makes them harder to find the root cause like in https://github.com/apache/airflow/issues/29025 ?

Comment (snjypl) [reply airflow/executors/kubernetes_executor.py:810]: @eladkal  i think, #29025  is more about the error  that we log around these part.  https://github.com/apache/airflow/blob/1e385ac36cf84cca92cc18bb528e381904b44783/airflow/executors/kubernetes_executor.py#L690-L714 .

These logs i believe are part of the scheduler logs and won't be visible as part of the task's log since we only fetch the logs from task's k8s pod in `kubernetes_executor.get_task_log`. 


regarding the exceptions, am not sure if i understand you correctly, but  i think, those exceptions are caught by the enclosing try/except and returned to the user.

**Review by potiuk — APPROVED** [review]
LGTM. @dstandish Maybe you can also take a look since you are working on this area (from different angle though - triggerers).

File: airflow/utils/log/file_task_handler.py
Comment (pierrejeambrun) [inline airflow/utils/log/file_task_handler.py]: Why do we need this check ? I think it only helps for custom executor that are not `BaseExecutor`, but other PR removed such check I believe.

**Review by pierrejeambrun — APPROVED** [review]
Small nit, otherwise LGTM

File: airflow/utils/log/file_task_handler.py
Comment (snjypl) [reply airflow/utils/log/file_task_handler.py]: thanks @pierrejeambrun  i went through the  discussion https://github.com/apache/airflow/issues/28276#issuecomment-1352590074 .  i have removed the `hasattr` check.

File: airflow/executors/kubernetes_executor.py
Comment (XD-DENG) [inline airflow/executors/kubernetes_executor.py:812]: Checking this part of code: why do we need to do the works above to get the pod name? The `ti.hostname` is just the pod name, isn't it?

cc @o-nikolas @snjypl

Comment (o-nikolas) [reply airflow/executors/kubernetes_executor.py:812]: This is not code that is new to this PR. It was just moved to a different location. If you see the `airflow/utils/log/file_task_handler.py` module, this code existed there before these changes.

**Pull request merged**
